### PR TITLE
Async achat put operation

### PR DESCRIPTION
**Pull request by xpbowler**
# Description

Currently, the memory in ChatEngine can only be synchronous even in the async methods like (achat, achat_stream).  
Fixes https://github.com/run-llama/llama_index/issues/15293

This PR adds class methods to several ABCs and an async version of calls for Azure database integration package.

`BaseMemory`: `aput`, with default implementation

`BaseChatStoreMemory`: `aput`

`BaseChatStore`: Async version of **all** class methods, with default implementations. This includes `delete()`, `put()`, `add()`, etc.

Children of `BaseChatStore`
- Only implemented async version for Azure at the moment. Others use default impl (call sync version)

## Type of Change

- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: llama-index-core/llama_index/core/storage/chat_store/base.py [hunk 1]
```diff
@@ -46,3 +46,31 @@ def delete_last_message(self, key: str) -> Optional[ChatMessage]:
...
+
+    async def aset_messages(self, key: str, messages: List[ChatMessage]) -> None:
+        """Async version of Get messages for a key."""
+        self.set_messages(key, messages)
+
+    async def aget_messages(self, key: str) -> List[ChatMessage]:
+        """Async version of Get messages for a key."""
+        return self.get_messages(key)
+
+    async def aadd_message(self, key: str, message: ChatMessage) -> None:
```
Comment (logan-markewich) [inline llama-index-core/llama_index/core/storage/chat_store/base.py:58]: small nit -- I would prefer `async def async_add_message(...)` -- the double `aa` makes it seem like a typo, and the vector store followed a similar pattern for naming the add function

File: llama-index-integrations/storage/chat_store/llama-index-storage-chat-store-azure/llama_index/storage/chat_store/azure/base.py
Comment (logan-markewich) [inline llama-index-integrations/storage/chat_store/llama-index-storage-chat-store-azure/llama_index/storage/chat_store/azure/base.py:1]: note to self: Will have to publish this AFTER core is published

File: llama-index-core/llama_index/core/storage/chat_store/base.py (hunk 1 above)
Comment (xpbowler) [reply llama-index-core/llama_index/core/storage/chat_store/base.py:58]: Sounds good

File: llama-index-legacy/llama_index/legacy/memory/chat_memory_buffer.py
Comment (logan-markewich) [inline llama-index-legacy/llama_index/legacy/memory/chat_memory_buffer.py:1]: no need to update legacy right?

**Review by logan-markewich — APPROVED** [review]

**Pull request merged**